### DISCUSSION TITLE
Templating configmap (or string in general) in tanka

### DISCUSSION THREAD

**Discussion by phongnguyen-monex** [grafana/tanka#635] [Q&A]
Jsonnet supports for deep merging is great for overriding values, but I haven't been able to find an up-to-date resource for templating string. For example, I'm trying to template a default loki deployment configmap for grafana agent with username and password:
```
# cat configmap.libsonnet
{
    apiVersion: 'v1',
    data: {
      'agent.yaml': |||
        loki:
            configs:
            - clients:
                - url: YOUR_LOKI_ENDPOINT
                    basic_auth:
                      username: YOUR_LOKI_USERNAME
                      password: YOUR_LOKI_PASSWORD
               # Other Loki configuration
      |||,
    },
    kind: 'ConfigMap',
    metadata: {
      name: 'grafana-agent-logs',
      namespace: 'YOUR_NAMESPACE',
    },
}
```
There are some methods for templating text file in the [old ksonnet documentation](https://ksonnet.io/docs/examples/template-text-files/) and current jsonnet documentation using Python-compatible string formatting with %, but it's rather unwieldy if the string have more than a few variables. I was wondering if there are better approach to string templating, or at least with structured yaml or json in configmap.

**Comment by strowi** [comment]
Also looking for a way to replace vars in data/text.
What would currently be the best way to handle this?

I am currently trying to do sth like the following, but it seems the $._config.realm is not available...

```
local config_file = importstr "config.txt";

{
  configmap: {
    apiVersion: 'v1',
    data: {
      'rabbitmq.conf': config_file % {
      "username": $._config.realm,
    },
    },
    kind: 'ConfigMap',
    metadata: {
      creationTimestamp: null,
      labels: {
        'io.kompose.service': 'rabbitmq',
      },
      name: 'rabbitmq-config',
    },
  }
}
```

**Comment by zephyros-dev** [comment]
For supported structured file format in jsonnet (like yaml, json or ini), I found that one can template string in them by using the std lib. For example: 
```
{
  yamlString: std.manifestYamlDoc(
    [
      {
        key: 'value',
      },
    ]
  ),
}
```

**Comment by strowi** [comment]
Oh after some time testing and debugging some formatting problems with json contained in yaml i came up with this one which seems easier for the eyes: ;)
This uses the k8s/ksonnet libs, but the replacment works the same.

```
#files/file1.yml
test1: %(var1)s,
test2: %(var2)s
```

```
{...
    ansibleData:: {
      'file.yml': (importstr 'files/file1.yml') % {
        var1: 'abc',
        varf2: $._config.var1,
      },
    },
    configMap:
      configMap.new('ansiblevars', $.ansible.ansibleData),
...}
```
